### PR TITLE
addDriverRunpath: init

### PR DESCRIPTION
**Pull request by jonringer**
## Description of changes

Less opinionated version of https://github.com/NixOS/nixpkgs/pull/196174

Allows for usage of `addDriverRunapth`, this way the setuphook is available to allow for a larger window for people to transition from `addOpenGLRunpath` setuphook.

Additional Context: [rfc #121](https://github.com/NixOS/rfcs/pull/121)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by jonringer** [comment]
should `cudaPackages.autoAddOpenGLRunpathHook` be promoted to top-level `pkgs`? I feel like it should

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -226,6 +226,12 @@ with pkgs;
 
   chkservice = callPackage ../tools/admin/chkservice { };
 
+  # addHardwareRunpath is the preferred pacakge name, as this enables
```
Comment (knedlsepp) [inline pkgs/top-level/all-packages.nix:229]: package

**Review by knedlsepp — APPROVED** [review]
Much more intuitive! 👍

File: nixos/doc/manual/release-notes/rl-2405.section.md [hunk 2]
```diff
@@ -26,4 +26,5 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 <!-- To avoid merge conflicts, consider adding your item at an arbitrary place in the list instead. -->
 
-- Create the first release note entry in this section!
+- `addHardwareRunpath` has been added to facilitate the deprecation of the old `addOpenGLRunpath` setuphook. This change is motivated by the desire to align usage of the setuphook with the evolving usages cases which now include all hardware acceleration.
```
Comment (jonringer) [inline nixos/doc/manual/release-notes/rl-2405.section.md:29]:
```suggestion
- `addHardwareRunpath` has been added to facilitate the deprecation of the old `addOpenGLRunpath` setuphook. This change is motivated by the desire to align usage of the setuphook with the evolving usage which now include all hardware acceleration.
```

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (jonringer) [inline pkgs/top-level/all-packages.nix:229]:
```suggestion
  # addHardwareRunpath is the preferred package name, as this enables
```

Comment (jonringer) [reply pkgs/top-level/all-packages.nix:229]: thanks :)

File: nixos/doc/manual/release-notes/rl-2405.section.md [hunk 3]
```diff
@@ -26,4 +26,5 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 <!-- To avoid merge conflicts, consider adding your item at an arbitrary place in the list instead. -->
 
-- Create the first release note entry in this section!
+- `addHardwareRunpath` has been added to facilitate the deprecation of the old `addOpenGLRunpath` setuphook. This change is motivated by the desire to align usage of the setuphook with the evolving usage which now include all hardware acceleration.
```
Comment (jonringer) [inline nixos/doc/manual/release-notes/rl-2405.section.md:29]:
```suggestion
- `addHardwareRunpath` has been added to facilitate the deprecation of the old `addOpenGLRunpath` setuphook. This change is motivated by the evolution of the setuphook to include all hardware acceleration.
```

**Comment by SomeoneSerge** [comment]
This is obviously an improvement.

Since we don't change the names too often, I wonder if we do want the new name to suggest that there should be one "merged" path for all of the impure dependencies: be that libgbm, or libcuda, or whatever is the next thing. This means that if we expose something impure to a program, we expose everything at once.

Btw the link to the rendered rfc is now broken

File: pkgs/build-support/add-hardware-runpath/default.nix [hunk 4]
```diff
@@ -0,0 +1,14 @@
+{ lib, stdenv }:
+
+stdenv.mkDerivation {
+  name = "add-hardware-runpath";
```
Comment (Atemu) [inline pkgs/build-support/add-hardware-runpath/default.nix:4]: Perhaps `add-driver-runpath` would be a more fitting name. We're not adding hardware into the runpath somehow but the driver thereof.

Comment (jonringer) [reply pkgs/build-support/add-hardware-runpath/default.nix:4]: that's fair

**Comment by jonringer** [comment]
changed the name to `addDriverRunpath`, as it is more aligned with "this is where the drivers are located" rather than "this is where the hardware is located"

File: pkgs/top-level/all-packages.nix [hunk 5]
```diff
@@ -226,6 +226,12 @@ with pkgs;
 
   chkservice = callPackage ../tools/admin/chkservice { };
 
+  # addDriverRunpath is the preferred package name, as this enables
+  # many more scenarios than just opengl now.
+  addDriverRunpath = callPackage ../build-support/add-driver-runpath { };
+
+  # addOpenGLRunpath should be added to aliases.nix after the 24.05 branch-off.
+  # Post 24.11 branch-off, this should throw an error in aliases.nix.
```
Comment (Atemu) [inline pkgs/top-level/all-packages.nix:234]: This is possible to do programmatically.

**Review by Atemu — APPROVED** [review]
Diff LGTM!

File: pkgs/build-support/add-driver-runpath/default.nix [hunk 6]
```diff
@@ -0,0 +1,14 @@
+{ lib, stdenv }:
+
+stdenv.mkDerivation {
+  name = "add-driver-runpath";
+
+  # Named "opengl-driver" for legacy reasons, but it is the path to
+  # hardware drivers installed by NixOS
```
Comment (samuela) [inline pkgs/build-support/add-driver-runpath/default.nix:7]: IIRC previously there was an initiative to renamed this path. perhaps that's worth revisiting at some point?

**Review by samuela — APPROVED** [review]
Looks great! Thank you so much for spearheading this @jonringer -- an important cleanup!

File: pkgs/top-level/all-packages.nix (hunk 5 above)
Comment (jonringer) [reply pkgs/top-level/all-packages.nix:234]: maybe, but I'm also a fan of just manually doing it at appropriate times.

**Comment by jonringer** [comment]
cc @infinisil before I start updating the Nixpkgs Manual entries, and doing a treewide rename.

**Comment by infinisil** [comment]
Just from a glance this looks fine to me :)

**Comment by Atemu** [comment]
Needs a rebase for the 24.05 manual, then we can merge :)

**Comment by 0david0mp** [comment]
hi guys, just a question for [PolyMC #1683](https://github.com/PolyMC/PolyMC/pull/1683). Is it correct? i've been trying to build it, but it would.  The problematic code:
```nix
...
    in
      ["--prefix POLLYMC_JAVA_PATHS : ${lib.makeSearchPath "bin/java" jdks}"]
      ++ lib.optionals stdenv.isLinux [
        "--set LD_LIBRARY_PATH ${addDriverRunpath.driverLink}/lib:${lib.makeLibraryPath runtimeLibs}"
        # xorg.xrandr needed for LWJGL [2.9.2, 3) https://github.com/LWJGL/lwjgl/issues/128
        "--prefix PATH : ${lib.makeBinPath runtimePrograms}"
      ];
...
```

throwing the next error:

```txt
error: evaluation aborted with the following error message: 'Function called without required argument "addDriverRunpath" at /nix/store/k00dg1wvmysx0sjss8rjqs07lbxqg9qn-source/nix/pkg/wrapper.nix:7'
```

I guess they only want a replacement for `addOpenGLRunpath.driverLink`, but the `addDriverRunpath` function needs another argument, is this correct?

if this is the case i would love to know any alternatives to the deprecated `addOpenGLRunpath.driverLink` approach. Thanks in advance.

**Comment by Atemu** [comment]
This looks more like the package wasn't provided the input. 

Make sure the Nixpkgs revision used actually includes this patch.

This is also off-topic for this PR. Please create a help thread in discourse or ask in Matrix instead.

**Comment by 0david0mp** [comment]
seems strange, the argument is declared (see [here](https://github.com/PolyMC/PolyMC/pull/1683/files)).  sorry for the inconvenience, im still learning the language and how to contribute to the community.  i will hide the comment.

it was an issue of the nixpkgs revision 😄, the `flake.lock` i was using on my home manager was different from the one i was using on the fork i used for building.

**Pull request merged**